ledger/participant-integration-api/src/main/scala/platform/store/ActiveLedgerStateManager.scala

```diff
@@ -69,146 +105,184 @@ private[platform] class ActiveLedgerStateManager[ALS <: ActiveLedgerState[ALS]](
+                      als.removeContract(ne.targetCoid)
+                    } else {
+                      als
+                    }),
+                    archivedIds =
+                      if (ne.consuming) acc.currentState.archivedIds + ne.targetCoid
+                      else acc.currentState.archivedIds,
+                  ),
+                  parties = acc.parties.union(nodeParties),
                 )
-                val gk = GlobalKey(nlkup.templateId, key)
-                val nodeParties = nlkup.key.maintainers
-
-                if (actAs.nonEmpty) {
-                  // If the submitter is known, look up the contract
-                  // Submitters being know means the transaction was submitted on this participant.
-                  val currentResult = acc.lookupContractByKey(gk)
-                  if (currentResult == nlkup.result) {
-                    ats.copy(
-                      parties = parties.union(nodeParties)
-                    )
-                  } else {
-                    ats.copy(
-                      errs = errs + Inconsistent(
-                        s"Contract key lookup with different results: expected [${nlkup.result}], actual [${currentResult}]"
-                      )
-                    )
-                  }
-                } else {
-                  // Otherwise, trust that the lookup was valid.
-                  // The submitter being unknown means the transaction was submitted on a different participant,
-                  // and (A) this participant may not know the authoritative answer to whether the key exists and
-                  // (B) this code is called from a Indexer and not from the sandbox ledger.
-                  ats.copy(
-                    parties = parties.union(nodeParties)
-                  )
-                }
+                (newState, true)
             }
-        }
-
-    val divulgedContractIds = divulgence -- st.archivedIds
+          },
+          rollbackBegin = (acc, _, _) => {
+            (acc.copy(rollbackStates = acc.currentState +: acc.rollbackStates), true)
+          },
+          leaf = (acc, nodeId, node) => handleLeaf(acc, nodeId, node),
+          exerciseEnd = (acc, _, _) => acc,
+          rollbackEnd = (acc, _, _) => {
+            val (beforeRollback, rest) = acc.rollbackStates match {
+              case Seq() => sys.error("IMPOSSIBLE: Rollback end but rollback stack is empty")
+              case head +: tail => (head, tail)
+            }
+            // Discard archives in the rollback but mark contracts created in the rollback as archived.
+            // This means that at the end of the traversal, we can use archivedIds to filter out
+            // both local and global contracts that are no longer active.
+            val newArchived =
+              beforeRollback.archivedIds union (acc.currentState.createdIds diff beforeRollback.createdIds)
+            acc.copy(
+              currentState = beforeRollback.copy(
```

Comment:
Why doesn't `rollbackEnd` restore the `currentState` to exactly as it was when pushed?
Why is any effect allowed to persist?

Reply:
Because we need `archiveIds` to contain all contracts that are not active at the moment including archived contracts and rolled back creates. The only purpose of `archivedIds` is to filter divulged contracts to those that are still active at the end.

Reply:
Isn't it the case that the ALS which contained the contracts created within the rollback, has already been dropped on the floor?

So what is it, that contains these contracts, which needs to be filtered?

Reply:
5 lines below, line 285. divulgence includes both local and global contracts.

Reply:
Tried to make it a bit clearer in c1c6f3eebbb70fa442678421fd27671cc4f5eab6

Reply:
Thanks.

So, to check I understand this correctly...

By _local_ contracts, you mean contracts created within a rollback?

The entry point to the `ActiveLedgerStateManager` code is a call to `addTransaction`, which is provided the `divulgence` (presumable already computed by an earlier transaction traversal, which includes local contracts created in a rollback node - why?).

So now `addTransaction` must re-track the _local_ contracts, just so it can remove them from the `divulgence`.

This seems very odd. Or am I still missing something?

Reply:
`local` = created within the transaction doesn’t matter if it was within a rollback or not and doesn’t matter if it was archived or not.
Divulgence is computed by `calculateBlindingInfo` which (as it should) also returns divulgence for local contracts (again regardless of whether they are created within a rollback or not).

`ActiveLedgerStateManager` wants divulgence for only those contracts that are active at the end of the transaction so it needs to trim 3 types of contracts:
1. Global (contracts created in a previous transaction) contracts that have been archived contracts.
2. Local contracts created outside of a rollback that have been archived in the same transaction.
3. Local contracts created in a rollback node (you can view them as having an implicit archive at the end of a rollback node).

The code here basically does the “implicit archive at the end of a rollback node”.

I agree it’s not super nice but I don’t really have a great idea for making it nicer without a pretty invasive change.

Reply:
Thanks for the explanation. The code is perfectly _nice_!

I'm just trying to understand where my mental model of _anything in a rollback node never happened_ breaks down. And it seems to be:

- 1: Regarding divulgence. That did/does happen.
- 2: Contracts created within rollbacks did get created, but then we archived them at the end.

Is that a fair summary?

Reply:
I think there are two points here:
1. First the model should be "rollbacks undo writes but they preserve reads". So a fetch in a rollback for example still happened and that matters for example for divulgence. While that’s related I don’t think that’s quite what makes the difference here.
2. The second point is about visibility/privacy. Regardless of whether something is rolled back, we still project in the rollback node. So if there was a create in a rollback node I might still get to see that. That’s why `BlindingInfo` includes nodes under rollback nodes.

The second point here is what matters. Here we only care about visibility of active contracts rather than visibility of the full transaction tree so we need to filter divulgence down to divulgence of active contracts. As mentioned above there are 3 typos of contracts that might no longer be active at the end of the transaction one of them being a create under a rollback node.

I think there might be two ways of making this nicer:
1. Expose activeness in blindinginfo.
2. Expose the list of all active contracts in `ALS` so we can filter it down

Both of those seemed to be more invasive than necessary so I left them out of this PR at least and tried to preserve the previous structure.

Reply:
Thanks again for more explanation. This is particular:
_"rollbacks undo writes but they preserve reads"._

Reply:
I’ve slightly simplified this after adding a helper method to `Transaction` which hopefully makes that less convoluted 3702149a568dbdfaec6bcd8e057eae80e1d261bb